stl/inc/utility

```diff
@@ -480,45 +480,46 @@ constexpr void swap(const pair<_Ty1, _Ty2>& _Left, const pair<_Ty1, _Ty2>& _Righ
 }
 #endif // _HAS_CXX23
 
-_EXPORT_STD template <class _Ty1, class _Ty2>
-_NODISCARD constexpr bool operator==(const pair<_Ty1, _Ty2>& _Left, const pair<_Ty1, _Ty2>& _Right) {
+_EXPORT_STD template <class _Ty1, class _Ty2, class _Uty1, class _Uty2>
+_NODISCARD constexpr bool operator==(const pair<_Ty1, _Ty2>& _Left, const pair<_Uty1, _Uty2>& _Right) {
     return _Left.first == _Right.first && _Left.second == _Right.second;
 }
 
 #ifdef __cpp_lib_concepts
-_EXPORT_STD template <class _Ty1, class _Ty2>
-_NODISCARD constexpr common_comparison_category_t<_Synth_three_way_result<_Ty1>, _Synth_three_way_result<_Ty2>>
-    operator<=>(const pair<_Ty1, _Ty2>& _Left, const pair<_Ty1, _Ty2>& _Right) {
+_EXPORT_STD template <class _Ty1, class _Ty2, class _Uty1, class _Uty2>
+_NODISCARD constexpr common_comparison_category_t<_Synth_three_way_result<_Ty1, _Uty1>,
+    _Synth_three_way_result<_Ty2, _Uty2>>
+    operator<=>(const pair<_Ty1, _Ty2>& _Left, const pair<_Uty1, _Uty2>& _Right) {
     if (auto _Result = _Synth_three_way{}(_Left.first, _Right.first); _Result != 0) {
         return _Result;
     }
     return _Synth_three_way{}(_Left.second, _Right.second);
 }
 #else // ^^^ defined(__cpp_lib_concepts) / !defined(__cpp_lib_concepts) vvv
 #if !_HAS_CXX20
-template <class _Ty1, class _Ty2>
-_NODISCARD constexpr bool operator!=(const pair<_Ty1, _Ty2>& _Left, const pair<_Ty1, _Ty2>& _Right) {
+template <class _Ty1, class _Ty2, class _Uty1, class _Uty2>
+_NODISCARD constexpr bool operator!=(const pair<_Ty1, _Ty2>& _Left, const pair<_Uty1, _Uty2>& _Right) {

```

Comment:
Do we want to make this change as a DR? It seems to me like the kind of change we shouldn't make backwards.

Reply:
I believe when I implemented one of my first LWG issue and did `ifdef` things @StephanTLavavej told me that LWG issues treated as DR. But maybe it was a different context.

cppreference also describes it as DR: https://en.cppreference.com/w/cpp/utility/pair/operator_cmp (the change added by https://en.cppreference.com/w/User:Fruderica , I'm not sure if that is @frederick-vs-ja or not)

Reply:
Yes, that's me.

Reply:
FYI, I have the same [libc++ patch](https://reviews.llvm.org/D144262) and @mordante wrote:
>We typically back-port LWG issues to older language versions.
However I'm a bit wary of retroactively applying this LWG issue, since it may affect existing code.
>Since `operator==` and `operator<=>` are added/changed in C++20, I feel less wary to change it there.
Doing this back to C++03 feels a bit scary.

Reply:
Our default position is that LWG issues represent defects in the library specification and we should fix that bug in all modes we support. We have the opposite default for proposals, since they usually represent feature additions. That said, the defaults are just defaults and there are sometimes good reasons to act differently.

I'm not concerned about this being breaking, but I may be missing something. I note that `pair<int, int>{} == {}` and `pair<int, int>{} == convertible_to_pair_int_int{}` are ill-formed in all language modes before and after this PR. Uses with explicit template argument lists will be broken (`std::operator==<int, int>(pair<int, int>{}, {})`) but that's severely cursed: we've happily broken code that calls operators by name before.

What potential breakage are you concerned about, @strega-nil-ms?

Reply:
```C++
struct BadPairs : std::pair<int, int>, std::pair<long, long> {};
static_assert(std::pair<int, int>{} == BadPairs{}, "");
```
This works before but becomes ambiguous after.

A (arguably) better fix is providing both the old and the new overloads. The old overloads are more specialized so the new overloads shouldn't introduce ambiguity.

Reply:
> This works before but becomes ambiguous after.

Thanks!
 
> A (arguably) better fix is providing both the old and the new overloads. The old overloads are more specialized so the new overloads shouldn't introduce ambiguity.

I'd like to avoid this if we can. If there's one thing `std` doesn't need more of, it's overloads of `operator==`. Making "good" code compile slower to avoid breaking a weird corner case is icky. I suppose my position here is still "I don't think this is a common pattern, I'd rather see what breaks before trying to fix it."

Reply:
I think we should implement this LWG issue resolution unconditionally. The example of multiply inheriting from `pair` is good to know about (thanks!) but it is extremely pathological and I don't expect to see that in the wild. I am even more strongly opposed than Casey is to providing both "old" and "new" overloads - not only would that impact throughput, but it has the potential to degrade compiler error messages, and might carry the risk of finding new and spicy compiler bugs.

There has been only one LWG issue that we've had to guard with a Standard mode (`operator<<(nullptr_t)`), and we did that only in response to finding significant actual breakage. (We can always revisit this decision if we encounter such breakage.)

Reply:
FWIW this is also the libc++ stance, and we'll align with MSVC STL on this by implementing this as a DR in all supported standard modes. We'll revisit that only if we find significant breakage in the wild.